{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Bioinformatics_#3.ipynb",
      "provenance": [],
      "collapsed_sections": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VJj2XYoE4UYo"
      },
      "source": [
        "# Repeating a few useful functions from the previous weeks of this course.\n",
        "\n",
        "* ReverseComplement(), finding the reverse complement of a gene.\n",
        "\n",
        "* HammingDistance(), finding the hamming  distance between two genes."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "5OEOJ49V4TLr"
      },
      "source": [
        "def ReverseComplement( P ):\n",
        "\n",
        "  dictionairy = { 'A': 'T', 'G' : 'C', 'T' : 'A', 'C' : 'G' }\n",
        "\n",
        "  P_r = P[::-1]\n",
        "\n",
        "  P_rc = \"\"\n",
        "\n",
        "  for i in range( len( P ) ):\n",
        "\n",
        "    P_rc += dictionairy[ P_r[ i ] ]\n",
        "\n",
        "  return P_rc\n",
        "\n",
        "\n",
        "\n",
        "def HammingDistance( fg, sg ):\n",
        "\n",
        "\n",
        "  distance = 0\n",
        "\n",
        "  for i in range( len( fg ) ):\n",
        "\n",
        "\n",
        "    if fg[ i ] != sg[ i ]:\n",
        "      distance += 1\n",
        "\n",
        "  return distance\n",
        "\n",
        "\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VFKSgFwP41KW"
      },
      "source": [
        "# Finding all possible DNA strings of length k and then finding all k-mers of hamming distance at most d within the genome that are present."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "dtZlXpmm41p9"
      },
      "source": [
        "\n",
        "genomes = input(\"Enter genome:\").split()\n",
        "\n",
        "d = 1\n",
        "\n",
        "k = 5\n",
        "\n",
        "kmer = \"\"\n",
        "\n",
        "freq_table = {}\n",
        "\n",
        "new_freq_table = {}\n",
        "\n",
        "alphabet = [ \"A\", \"T\", \"C\", \"G\" ]\n",
        "\n",
        "dictionary = [\"A\" * k]\n",
        "\n",
        "num = list( \"0\" * k )\n",
        "\n",
        "len_alpha = len( alphabet )\n",
        "\n",
        "g_length = len( genomes[ 0 ] )\n",
        "\n",
        "\n",
        "\n",
        "def add( num, position ):\n",
        "\n",
        "  if int( num[ position ] ) + 1 > len_alpha - 1:\n",
        "\n",
        "    if position + 1 > k-1:\n",
        "      return\n",
        "\n",
        "    num[ position ] = '0'\n",
        "\n",
        "    add( num, position + 1 )\n",
        "    return\n",
        "\n",
        "  num[ position ] = str( int( num[ position ] ) + 1  )\n",
        "\n",
        "  kmer = []\n",
        "\n",
        "  for i in range( len( num ) ):\n",
        "\n",
        "    kmer.append( alphabet[int(num[i])] )\n",
        "\n",
        "  dictionary.append( \"\".join(kmer) )\n",
        "  \n",
        "\n",
        "def find_kmers( genome ):\n",
        "  for i in range( len_alpha ** k ):\n",
        "\n",
        "    if int( num[ 0 ] ) + 1 > len_alpha - 1:\n",
        "\n",
        "      num[ 0 ] = '0'\n",
        "\n",
        "      add( num, 1 )\n",
        "    \n",
        "    else:\n",
        "\n",
        "      num[ 0 ] = str( int( num[ 0 ] ) + 1  )\n",
        "\n",
        "      kmer = []\n",
        "\n",
        "      for i in range( len( num ) ):\n",
        "\n",
        "        kmer.append( alphabet[int(num[i])] )\n",
        "\n",
        "      dictionary.append( \"\".join(kmer) )\n",
        "\n",
        "\n",
        "\n",
        "  for i in range( g_length - k + 1 ):\n",
        "\n",
        "\n",
        "    # Add the k-mer if it is not in the frequency table.\n",
        "    if genome[ i:(i + k) ] not in freq_table.keys():\n",
        "\n",
        "      freq_table[ genome[ i:(i + k) ] ] = 1\n",
        "\n",
        "    else:\n",
        "\n",
        "      freq_table[ genome[ i:(i + k) ] ] += 1\n",
        "\n",
        "  count = 0\n",
        "\n",
        "\n",
        "  for kmer in dictionary:\n",
        "\n",
        "    for key in freq_table.keys():\n",
        "\n",
        "      if HammingDistance( kmer, key ) <= d:\n",
        "\n",
        "        count += freq_table[ key ]\n",
        "\n",
        "      #if HammingDistance( ReverseComplement( kmer ), key ) <= d:\n",
        "\n",
        "        #count += freq_table[ key ]\n",
        "\n",
        "    new_freq_table[ kmer ] = count\n",
        "\n",
        "    count = 0\n",
        "\n",
        "\n",
        "  buff = {}\n",
        "\n",
        "  for x, y in new_freq_table.items():\n",
        "\n",
        "    if y != 0:\n",
        "\n",
        "      buff[ x ] = y\n",
        "\n",
        "  return buff\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "new_freq_table = find_kmers( genomes[ 3 ] )\n",
        "\n",
        "for kmer in new_freq_table.keys():\n",
        "\n",
        "  kmer_present = 0\n",
        "\n",
        "  for genes in genomes:\n",
        "\n",
        "    #print( genes, kmer )\n",
        "\n",
        "    for i in range( len( genes ) - k + 1 ):\n",
        "\n",
        "      if HammingDistance( kmer, genes[ i:( i + k ) ] ) <= d:\n",
        "\n",
        "        kmer_present += 1\n",
        "        break\n",
        "\n",
        "\n",
        "\n",
        "  if kmer_present >= len( genomes ) :\n",
        "\n",
        "    print( kmer )\n",
        "\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "BQCu91gA7NDt"
      },
      "source": [
        "# Finding the profile matrix of a genome."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Bg3XDm1f7SNr"
      },
      "source": [
        "\n",
        "kmer = list( input(\"Enter Genome: \") )\n",
        "\n",
        "#kmer = list( \"ACCTGTTTATTGCCTAAGTTCCGAACAAACCCAATATAGCCCGAGGGCCT\" )\n",
        "\n",
        "#profile = [['0.2', '0.2', '0.3', '0.2', '0.3'], ['0.4', '0.3', '0.1', '0.5', '0.1'], ['0.3', '0.3', '0.5', '0.2', '0.4'], ['0.1', '0.2', '0.1', '0.1', '0.2']]\n",
        "\n",
        "profile = [ ]\n",
        "\n",
        "\n",
        "for i in range(4):\n",
        "\n",
        "  print( \"Input The \", i, \"th row of the profile matrix: \")\n",
        "\n",
        "  profile.append( input().split() )\n",
        "\n",
        "\n",
        "k = int( input(\"Enter k: \") )\n",
        "\n",
        "#k = 5\n",
        "\n",
        "dictionary = { \"A\": 0, \"C\": 1, \"G\": 2, \"T\": 3 }\n",
        "\n",
        "prob_dict = {}\n",
        "\n",
        "for i in range( len(kmer) - k + 1 ):\n",
        "\n",
        "\n",
        "  prob = 1\n",
        "\n",
        "  for j in range( k ):\n",
        "\n",
        "\n",
        "    #print( (kmer[ i:(i + k) ][j]), profile[ dictionary[ (kmer[ i:(i + k) ][j]) ]  ], float( profile[ dictionary[ (kmer[ i:(i + k) ][j]) ]  ] [ j ] ))\n",
        "\n",
        "    prob *= float( profile[ dictionary[ (kmer[ i:(i + k) ][j]) ]  ] [ j ] )\n",
        "\n",
        "\n",
        "  prob_dict[ \"\".join( kmer[ i:(i + k) ] ) ] = prob\n",
        "\n",
        "\n",
        "print( max( prob_dict, key=prob_dict.get ) )"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "odVD3n927Yye"
      },
      "source": [
        "## The Hamming distance function for reference again."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "tkAUTkn77cXb"
      },
      "source": [
        "def HammingDistanceArray( fg, sg ):\n",
        "\n",
        "\n",
        "  distance = 0\n",
        "\n",
        "  for i in range( len( fg ) ):\n",
        "\n",
        "    for j in range( len(fg[0]) ):\n",
        "\n",
        "      \n",
        "      if list( fg[ i ] ) [ j ] != sg[ j ]:\n",
        "        distance += 1\n",
        "\n",
        "  return distance\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Wdbty-wU7jUs"
      },
      "source": [
        "# This code brute forces a motif from a genome.\n",
        "\n",
        "* Initially selects a motif from the first k-length gene of each string.\n",
        "\n",
        "* Scores this motif and then replaces each gene in each segment of the motif to get a better scoring motif.\n",
        "\n",
        "* Outputs the overall best motif from it's search."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "CPb4uaUG7k2j"
      },
      "source": [
        "\n",
        "\n",
        "DNA = input(\"DNA strings: \").split()\n",
        "\n",
        "k = int( input(\"k: \") )\n",
        "\n",
        "initial_motif_matrix = [ string[0:k] for string in DNA ]\n",
        "\n",
        "len_DNA = len(DNA)\n",
        "\n",
        "alphabet = [ \"A\", \"C\", \"G\", \"T\"]\n",
        "\n",
        "alpha_dict = { \"A\": 0, \"C\": 1, \"G\": 2, \"T\": 3 }\n",
        "\n",
        "motif = []\n",
        "\n",
        "best_motif = []\n",
        "\n",
        "motif_score = 0\n",
        "\n",
        "best_ham = 0\n",
        "\n",
        "\n",
        "\n",
        "def Score_Motif( motifs, consensus_string ):\n",
        "\n",
        "  score_motif = 0\n",
        "\n",
        "  for kmer in motif:\n",
        "\n",
        "    for i in range( k ):\n",
        "\n",
        "      if kmer[i] != consensus_string[i]:\n",
        "\n",
        "        score_motif += 1\n",
        "\n",
        "  return score_motif\n",
        "\n",
        "\n",
        "\n",
        "def Find_Profile( motifs, position ):\n",
        "\n",
        "\n",
        "  for i in range( len_DNA - position ):\n",
        "\n",
        "    freq_table = {}\n",
        "\n",
        "    for j in range( 1, len( DNA[0] ) - k + 1):\n",
        "\n",
        "      motif_i_trial = DNA[ position + i ][ j:(j + k) ]\n",
        "\n",
        "      freq_table[ motif_i_trial ] = HammingDistanceArray( motif[0:position], motif_i_trial )\n",
        "\n",
        "    motif[ i + position ] = min( freq_table, key=freq_table.get )\n",
        "\n",
        "\n",
        "def Find_Profile2( motifs, position ):\n",
        "\n",
        "\n",
        "  # Generate a blank profile matrix.\n",
        "  profile = [ ]\n",
        "\n",
        "  consensus_string = \"\"\n",
        "\n",
        "  consensus_prob = []\n",
        "\n",
        "  for nuc in alphabet:\n",
        "\n",
        "    blank_kmer = [ 0 for i in range( k ) ]\n",
        "\n",
        "    for mot in motifs[ 0:position ]:\n",
        "\n",
        "      for j in range( k ):\n",
        "\n",
        "        if mot[ j ] == nuc:\n",
        "\n",
        "          blank_kmer[ j ] += 1\n",
        "\n",
        "\n",
        "    profile.append( blank_kmer )\n",
        "\n",
        "\n",
        "  for i in range( k ):\n",
        "\n",
        "    total_freq = profile[ 0 ][ i ] + profile[ 1 ][ i ] + profile[ 2 ][ i ] + profile[ 3 ][ i ]\n",
        "\n",
        "    for j in range( 4 ):\n",
        "\n",
        "      profile[ j ][ i ] = profile[ j ][ i ] / total_freq\n",
        "\n",
        "      consensus_prob.append( profile[ j ][ i ] )\n",
        "\n",
        "    consensus_string += list(alpha_dict.keys())[consensus_prob.index( max( consensus_prob ) )]\n",
        "    consensus_prob = []\n",
        "  \n",
        "\n",
        "  # Need to check whether this piece of code works.\n",
        "  for i in range( len_DNA - position ):\n",
        "\n",
        "    freq_table = {}\n",
        "\n",
        "    for j in range( 0, len( DNA[0] ) - k + 1):\n",
        "\n",
        "      motif_i_trial = DNA[ position + i ][ j:(j + k) ]\n",
        "\n",
        "      prob = 1\n",
        "\n",
        "      for p in range( k ):\n",
        "\n",
        "        # There may be an error here. \n",
        "        prob *= profile[ alpha_dict[ motif_i_trial[ p ] ]  ] [ p ]\n",
        "\n",
        "      freq_table[ motif_i_trial ] = 1 - prob\n",
        "\n",
        "    motif[ i + position ] = min( freq_table, key=freq_table.get )\n",
        "\n",
        "\n",
        "  return consensus_string\n",
        "\n",
        "\n",
        "\n",
        "# Crafting the initial motif matrix.\n",
        "motif = initial_motif_matrix\n",
        "\n",
        "#best_ham = HammingDistanceArray( motif[ 0:len_DNA ], motif[ 0 ] )\n",
        "\n",
        "\n",
        "profile = [ ]\n",
        "\n",
        "consensus_string = \"\"\n",
        "\n",
        "consensus_prob = []\n",
        "\n",
        "for nuc in alphabet:\n",
        "\n",
        "  blank_kmer = [ 0 for i in range( k ) ]\n",
        "\n",
        "  for mot in motif:\n",
        "\n",
        "    for j in range( k ):\n",
        "\n",
        "      if mot[ j ] == nuc:\n",
        "\n",
        "        blank_kmer[ j ] += 1\n",
        "\n",
        "\n",
        "  profile.append( blank_kmer )\n",
        "\n",
        "\n",
        "for i in range( k ):\n",
        "\n",
        "  total_freq = profile[ 0 ][ i ] + profile[ 1 ][ i ] + profile[ 2 ][ i ] + profile[ 3 ][ i ]\n",
        "\n",
        "  #print( 'Profile matrix before modification: ', profile )\n",
        "\n",
        "  for j in range( 4 ):\n",
        "\n",
        "    profile[ j ][ i ] = profile[ j ][ i ] / total_freq\n",
        "\n",
        "    consensus_prob.append( profile[ j ][ i ] )\n",
        "\n",
        "  #print( 'Profile matrix after modification: ', profile )\n",
        "\n",
        "  #consensus_string += list(alpha_dict.keys())[list(alpha_dict.values()).index( max( consensus_prob ) )]\n",
        "\n",
        "  #consensus_string += list(alpha_dict.keys())[consensus_prob.index( max( consensus_prob ) )]\n",
        "  consensus_string += alphabet[consensus_prob.index( max( consensus_prob ) )]\n",
        "  consensus_prob = []\n",
        "\n",
        "  #print( consensus_string )\n",
        "\n",
        "print( 'profile matrix: ', profile )\n",
        "\n",
        "best_score = Score_Motif( motif, consensus_string )\n",
        "\n",
        "print( \"Initial motif: \", motif )\n",
        "\n",
        "\n",
        "for j in range( 1, (len( DNA[0] ) - k + 1) ):\n",
        "\n",
        "  motif[ 0 ] = DNA[ 0 ][ j:(j + k) ]\n",
        "\n",
        "  #print( motif[ 0 ] )\n",
        "\n",
        "  #print( DNA[0], DNA[ 0 ][ j:(j + k) ] )\n",
        "\n",
        "  for i in range( 1, len_DNA ):\n",
        "\n",
        "    # I don't understand whats going on here.\n",
        "    consensus_string = Find_Profile2( motif, i )\n",
        "    \n",
        "  Motif_score = Score_Motif( motif, consensus_string )\n",
        "  #print( \"\\n\", motif, \" \", Motif_score, \" \", consensus_string )\n",
        "\n",
        "  if best_score > Motif_score:\n",
        "\n",
        "    best_motif = motif.copy()\n",
        "    best_score = Motif_score\n",
        "\n",
        "  #print( \"\\n Motif: \", motif )\n",
        "\n",
        "  #print( \"\\n   Best motif: \", best_motif )\n",
        "\n",
        "print( best_motif )\n",
        "\n",
        "\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "uXWDqo5e9LqE"
      },
      "source": [
        "# An test input to the previous function.\n",
        "GGCGTTCAGGCA AAGAATCAGTCA CAAGGAGTTCGC CACGTCAATCAC CAATAATATTCG\n",
        "3"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "kaymbFV69RJc"
      },
      "source": [
        "# Outputting the results of the previous algorithm, but separated by newlines\n",
        "# as opposed to spaces.\n",
        "for string in best_motif:\n",
        "\n",
        "  print( string )"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "aY-2NTPZ9kbj"
      },
      "source": [
        "# Making modifications to the previous block as I found some bugs."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "cNzqVPBv9sia"
      },
      "source": [
        "DNA = input(\"DNA strings: \").split()\n",
        "\n",
        "k = int( input(\"k: \") )\n",
        "\n",
        "initial_motif_matrix = [ string[0:k] for string in DNA ]\n",
        "\n",
        "len_DNA = len(DNA)\n",
        "\n",
        "alphabet = [ \"A\", \"C\", \"G\", \"T\"]\n",
        "\n",
        "alpha_dict = { \"A\": 0, \"C\": 1, \"G\": 2, \"T\": 3 }\n",
        "\n",
        "motif = []\n",
        "\n",
        "best_motif = []\n",
        "\n",
        "motif_score = 0\n",
        "\n",
        "best_ham = 0\n",
        "\n",
        "\n",
        "\n",
        "def Score_Motif( motifs, consensus_string ):\n",
        "\n",
        "  score_motif = 0\n",
        "\n",
        "  for kmer in motif:\n",
        "\n",
        "    for i in range( k ):\n",
        "\n",
        "      if kmer[i] != consensus_string[i]:\n",
        "\n",
        "        score_motif += 1\n",
        "\n",
        "  return score_motif\n",
        "\n",
        "\n",
        "\n",
        "def Find_Profile( motifs, position ):\n",
        "\n",
        "\n",
        "  for i in range( len_DNA - position ):\n",
        "\n",
        "    freq_table = {}\n",
        "\n",
        "    for j in range( 1, len( DNA[0] ) - k + 1):\n",
        "\n",
        "      motif_i_trial = DNA[ position + i ][ j:(j + k) ]\n",
        "\n",
        "      freq_table[ motif_i_trial ] = HammingDistanceArray( motif[0:position], motif_i_trial )\n",
        "\n",
        "    motif[ i + position ] = min( freq_table, key=freq_table.get )\n",
        "\n",
        "\n",
        "def Find_Profile2( motifs, position ):\n",
        "\n",
        "\n",
        "  # Generate a blank profile matrix.\n",
        "  profile = [ ]\n",
        "\n",
        "  consensus_string = \"\"\n",
        "\n",
        "  consensus_prob = []\n",
        "\n",
        "  for nuc in alphabet:\n",
        "\n",
        "    blank_kmer = [ 1 for i in range( k ) ]\n",
        "\n",
        "    for mot in motifs[ 0:position ]:\n",
        "\n",
        "      for j in range( k ):\n",
        "\n",
        "        if mot[ j ] == nuc:\n",
        "\n",
        "          blank_kmer[ j ] += 1\n",
        "\n",
        "    profile.append( blank_kmer )\n",
        "\n",
        "\n",
        "  for i in range( k ):\n",
        "\n",
        "    total_freq = profile[ 0 ][ i ] + profile[ 1 ][ i ] + profile[ 2 ][ i ] + profile[ 3 ][ i ]\n",
        "\n",
        "    for j in range( 4 ):\n",
        "\n",
        "      profile[ j ][ i ] = profile[ j ][ i ] / total_freq\n",
        "\n",
        "      consensus_prob.append( profile[ j ][ i ] )\n",
        "\n",
        "    consensus_string += list(alpha_dict.keys())[consensus_prob.index( max( consensus_prob ) )]\n",
        "    consensus_prob = []\n",
        "  \n",
        "\n",
        "  # Need to check whether this piece of code works.\n",
        "  for i in range( len_DNA - position ):\n",
        "\n",
        "    freq_table = {}\n",
        "\n",
        "    for j in range( 0, len( DNA[0] ) - k + 1):\n",
        "\n",
        "      motif_i_trial = DNA[ position + i ][ j:(j + k) ]\n",
        "\n",
        "      prob = 1\n",
        "\n",
        "      for p in range( k ):\n",
        "\n",
        "        # There may be an error here. \n",
        "        prob *= profile[ alpha_dict[ motif_i_trial[ p ] ]  ] [ p ]\n",
        "\n",
        "      freq_table[ motif_i_trial ] = 1 - prob\n",
        "\n",
        "    motif[ i + position ] = min( freq_table, key=freq_table.get )\n",
        "\n",
        "\n",
        "  return consensus_string\n",
        "\n",
        "\n",
        "\n",
        "# Crafting the initial motif matrix.\n",
        "motif = initial_motif_matrix\n",
        "\n",
        "#best_ham = HammingDistanceArray( motif[ 0:len_DNA ], motif[ 0 ] )\n",
        "\n",
        "\n",
        "profile = [ ]\n",
        "\n",
        "consensus_string = \"\"\n",
        "\n",
        "consensus_prob = []\n",
        "\n",
        "for nuc in alphabet:\n",
        "\n",
        "  blank_kmer = [ 1 for i in range( k ) ]\n",
        "\n",
        "  for mot in motif:\n",
        "\n",
        "    for j in range( k ):\n",
        "\n",
        "      if mot[ j ] == nuc:\n",
        "\n",
        "        blank_kmer[ j ] += 1\n",
        "\n",
        "  profile.append( blank_kmer )\n",
        "\n",
        "\n",
        "for i in range( k ):\n",
        "\n",
        "  total_freq = profile[ 0 ][ i ] + profile[ 1 ][ i ] + profile[ 2 ][ i ] + profile[ 3 ][ i ]\n",
        "\n",
        "  #print( 'Profile matrix before modification: ', profile )\n",
        "\n",
        "  for j in range( 4 ):\n",
        "\n",
        "    profile[ j ][ i ] = profile[ j ][ i ] / total_freq\n",
        "\n",
        "    consensus_prob.append( profile[ j ][ i ] )\n",
        "\n",
        "  #print( 'Profile matrix after modification: ', profile )\n",
        "\n",
        "  #consensus_string += list(alpha_dict.keys())[list(alpha_dict.values()).index( max( consensus_prob ) )]\n",
        "\n",
        "  #consensus_string += list(alpha_dict.keys())[consensus_prob.index( max( consensus_prob ) )]\n",
        "  consensus_string += alphabet[consensus_prob.index( max( consensus_prob ) )]\n",
        "  consensus_prob = []\n",
        "\n",
        "  #print( consensus_string )\n",
        "\n",
        "print( 'profile matrix: ', profile )\n",
        "\n",
        "best_score = Score_Motif( motif, consensus_string )\n",
        "\n",
        "print( \"Initial motif: \", motif )\n",
        "\n",
        "\n",
        "for j in range( 1, (len( DNA[0] ) - k + 1) ):\n",
        "\n",
        "  motif[ 0 ] = DNA[ 0 ][ j:(j + k) ]\n",
        "\n",
        "  #print( motif[ 0 ] )\n",
        "\n",
        "  #print( DNA[0], DNA[ 0 ][ j:(j + k) ] )\n",
        "\n",
        "  for i in range( 1, len_DNA ):\n",
        "\n",
        "    # I don't understand whats going on here.\n",
        "    consensus_string = Find_Profile2( motif, i )\n",
        "    \n",
        "  Motif_score = Score_Motif( motif, consensus_string )\n",
        "  #print( \"\\n\", motif, \" \", Motif_score, \" \", consensus_string )\n",
        "\n",
        "  if best_score > Motif_score:\n",
        "\n",
        "    best_motif = motif.copy()\n",
        "    best_score = Motif_score\n",
        "\n",
        "  #print( \"\\n Motif: \", motif )\n",
        "\n",
        "  #print( \"\\n   Best motif: \", best_motif )\n",
        "\n",
        "print( best_motif )"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "sEofaKmT9vBz"
      },
      "source": [
        "# Again outputting the best motif separated by newlines as opposed to spaces.\n",
        "for string in best_motif:\n",
        "\n",
        "  print( string )"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "RfgoJmp199y6"
      },
      "source": [
        "# An extension of the Hammign distance function."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "bdx0e-sv9-tj"
      },
      "source": [
        "def HammingDistanceArray( fg, sg ):\n",
        "\n",
        "\n",
        "  distance = 0\n",
        "\n",
        "  for i in range( len( fg ) ):\n",
        "\n",
        "    if fg[ i ] != sg[ i ]:\n",
        "      distance += 1\n",
        "\n",
        "  return distance"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "NsH-QtG2-KUi"
      },
      "source": [
        "\n",
        "def DistanceBetweenPatternAndStrings(Pattern, Dna):\n",
        "\n",
        "  Dna = Dna.split()\n",
        "\n",
        "  k = len( Pattern )\n",
        "  distance_string = 0\n",
        "  distance = 0\n",
        "\n",
        "  for string in Dna:\n",
        "\n",
        "    #print( string )\n",
        "\n",
        "    distance_string = HammingDistanceArray( Pattern, string[0:k] )\n",
        "  \n",
        "    #print( 'distance_string initial: ', distance_string )\n",
        "\n",
        "    #print( len(string) - k + 1 )\n",
        "\n",
        "    for i in range( len( string ) - k + 1 ):\n",
        "      \n",
        "\n",
        "      #print( string[ i:(i + k)] )\n",
        "\n",
        "      #print( 'AAA', string[ i:(i + k)], HammingDistanceArray( Pattern, string[ i:(i + k)] ) )\n",
        "\n",
        "      if distance_string > HammingDistanceArray( Pattern, string[ i:(i + k)] ):\n",
        "\n",
        "        distance_string = HammingDistanceArray( Pattern, string[ i:(i + k)] )\n",
        "\n",
        "\n",
        "    #print( 'distance_string then:', distance_string )\n",
        "\n",
        "    distance += distance_string\n",
        "\n",
        "  print( distance )\n",
        "\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-eIM41FN-Mbb"
      },
      "source": [
        "DistanceBetweenPatternAndStrings( 'GAGTACA', 'CCAGCGCTGCTTTATTGTTAAGCCGTTATCCTCCAAGTTGAGTCCTCGCGTAGCTCTAGAACGTCGGATCGGCCCACAACCAACAAACTAAAAATCAGGCGAGATTGCA TGAGTGGACAAACGGTTATGAAATCTGTATCCTCAATCTTAGGTGGACTCTCGCGCAGAGCGCGTGTACGGAGAGCTGGTCCCGGTGACCAAAGATCGTTCAGATTCGT GAGAGCGTTGCCGAGTAGCCCCATACCAACATGGTTCCACGGATCGTTTATCCCCTCGGACCTGGTTCCTGTCCACGCTATTTCCAAACGCTAGATATTTCGAGTCTAG ATTCTCAACGCAGTCGGCCCAAGTGCGGTCATACGGCATGAAGTATCATCCAATAAAAGTAAACGCCTCATATATCCATAGAGTACACAATCCGAGTAAATACCAGTCC CTTCCCGTCCACATCCTGTACCTGAGGCACTGGGCTGCCATAGATAAAAAAAGGGAACTAACGATTAGACCGTTGCCTGGATGGCCGGACGTCTCTTATCCCGCCACAT TCGGAAGTTTAAGGCAACGTCTTCGCGACGATAACGCTCACAACAGGCGCGCACCACAGGCCGGACTGGTATATGACAGGCTTCGGATGGATGATGTCGATATCGTATG ACGTAGTTCGATATATCACCTCGACCCTATGCGGTCTTGAGATCAGCAGGCTATGGCGATCAACAAACGCGGAAGCACCGGGGTACAGCCGGACTTGGTCTGCTAGATC TCGGGACTCATCTCAGCTAATACGCGATCTTGGTTGTGGTTTATCGTTGGGAGCAGAGCCCCCCGAGTTCAAGGTTCTGGGTTAGATACGATTCTCCCGCCAGGTCGTG AATTAAATTGGGCCCGAAGCTAAGATTTTCCGGGGAGCGCCGGTCCAGGATCTGTGACTCAAAACATACAACCCCAGTCAGCCTTAGCCGAAGTGGTCTCCCTCCGGGC TCGAAGAGAAGCCTCATCATCAGCGACCGTTCCTAATTTAGTAGGAAATTCGCAAATTGCGACGCAGTATAACCTCTAAGCCTGGACCTTCGATGCGGAAACGAAAGGG CGGCATTGGCGTTTACTATTATTTCTTTTAGCTTCGCTTTAATTCTGGTGTAACATTCCCGATACGCGATCTCATTCCTGTCGCGGTGGAGTCCTGGTACGCAATGATC CTTGGTGAATCGTGCCTCTAGCGGCAGTGGTTCGACGCTTCGGCAAACTTCAGGCCGTTCTAGTACGCTCCTTCGTAACTGCGACCACCCCCTTGATGCGAGAGCTGTT CTATTGACGAGCGATGGGAGCACATAAAAGAGGTTTGGTATATGCCCGCCGATTAGCACCTTAGGGCAGAGGACGTTGGGCAAAATTCTGGCATCTAAGTGTTAGGCTA AAAGAAAGTAAGGCCCGAGACGACGTTACGTCAGCGTAACGGCCATACGCTAGATGTCATTAAACAGTATTGTACCCGACGGTTTATCTTGGCTCATACGCTTTTAACG ATTATCCTCATAGTAATTCAATGCATTCATCCACTGATAGAAAATTATAACCTCCCTATGGAGCCCTTCCGGCACCCTTTGGGCTATTGCGAGACTGTCCGCTAGTAAC GTTTTCGATTGCAGGCTGTGCGTCCTTAGCGCGGGCAGTTAATATAAAGGGATTTGTTTCCTACCTCTTATATGCCGATAGCACAGATCAACGAATGAAAGACCAAATT AGGGGTCGCCTCTTACCTGCGCCCCTCTGGAGCCCTTTATAGACTTTCTAAGCAAATGCTGATAACTCGTGTGCGTAGAGTTCTGAATAGGCAAACAAGGTACTATAAG CCGAACTCCATCCAAACAAAATTGCTTCACGTCGCAGGATAGATGACGCAAATGAGGCAACTTGTTCGTATAACGGTTGTCGTACGCATCATTCGCCACATAAAGTACA AACCTGACCTCAGAGTTCCGCCATGTTTTTCCCCGAACCAGCGACATCGCATCCTGTCTTTTCGCGGGAAATGCCAACTAGAGCGGCATGACTGAGGGCAGTGGATCGG GCCATAGTGTCCATATATGACACAGGTCGTTTTTTGGAGTTTGTCGGCAACAAAGGACTGATTATCTGTACAGTTAAACGCCCCATCGATCGAGGGCCTGTAATTTTCT GCACATGATCCACGGATCAATTGTAACTGGGGTAAAGAAGCTGCAGGTCAAAAGGTTCGCTTGACTAAATGCCGGATAGCGCCAAGCCAGGCTCTCTAGGGTTCTTCAC CAGCGGTACGGAAACGCCTCAAGCGGGTGGATATGGCCGCGGCCACACCAACCAGGGGCGCCCATGAAGCGGAAGGCTTTCTATCTAGCATTCTGGAATAGTCCATATG TGATTCATTCCCGAATACTCCAAGTGGGAGAATGGTCCTCCTGGGGTGTGTACTACCGCGTGAATAAGTACTTCCGAGTAATACACCTCCGAATTTTGATATTAATGGA TGTGGTGGTAACGTAAAACGGATAAAGCATGCACACACTAACGGCCTAGTCTGGATTCCACGCCCAGGAATCTTCAGGCACATTAGGATGATGCCTTTTAATATCTACG TCGGAGATGGAAGCTACAGGTCTATTAGACCCTGTAAGGTCGCGTGCCCCATCGCAAATAACCGCGCAGGCTGCCCACGCCATTGTAGTGTTTCTGCGCGCCAGCAGAT GATGCTCCCACTGGTCCAAAAAAACTGTAGTGAATACCCCCCTGACACCGGCGCCTGGACATGATTCGGCGACGGTTGCTCTGCTGCCTATGAGTTTCCTGCGAATACG CGTAATCGGAAAGTGCGTCCAGAAGCCTCTGCGGGCTCTACTGTCGCAGTCATGTCTCAAAGAAAAGACGCATTCCAAACCATTCAATCTGACACGTCATCACTTACGT ACACTCCCATATGAAGAGTCGTGCCGCACTAGGACGCGTGTGACAAACCCTTGGTAAGCTGGCAGAGGAGAACCATTAGTCTTGAGCGGTTTTCCCGAACCATCTAGGT TCACCGTTAATGCAGCGCCGATCGAGTTTGGACCACTCGTCTTGACACCTTCATCGTTACGTGCGACATGGAACGCTAGCGGCGTAAAGAGGTAGTTGGGACGGTGTGT AAAATTGGTTAAATACGATCAGGCTACAGCCCTGGGCGGCCGCATCCATATGACTGCTCCCATACTATAAGTCGATTACAGGGCACACGGTTTATGAATTTGCAGTTGC CGGGAAGGCTAAGTGAAAGTACCTGCCCTGCAGCGTGTTACTAGTCTGGCACACTGGGTTTGCCGCGCTACATGCCATGGAGTGATCCCTGAGCCACCAATGTCGGCCA TATGCCCGGAGTCATAGTAGGTGAGGCCTGTGTCGGACTCACTCGGGTCCATTCACGTACAACTCGGACGTTGATAGAAAAGCCAATGCAGCGTGACCAGGCAATGACG GCTTAATGAGAGCTCAGCGATGCCTCGTCTAGAACCGCATATCCGTGCCAAACAATCAATTAGTTTCGTAACAGTCACCTCACTATTGCCTATTGGTCAGATGAGTCGA CGGGGCCAAGGAAGGTATGCGTCGGATTACAAAATCCCTGGAACGGGACTCTCTAAATACCTGGACGGGGACCCTACATTAAATATTAAAATAGACTATACCTACTGAA AGAAATGAATCAAGGCTCGGTGTGTTACGCGTCGATTGCAACGTCCCTTAAGCCTTTACGTACCGGGGGAGGTTTGGGGTAGTGCGACTTAGGGCTATGCACGACTCCA CAATTCTTAATGTGCGGATAAGACTGTTAAACATGCCGTTGGAATCTGGAAGGCTCCATCGGCTCCCTGCCTACCTAGCAAACCTCTAATTATCGGAACTGTTTGAGAC AGTAGCGCGCCTGAACTCAAAGGAAGTGAACACACGTAGCACGAGAACGAGAGATGTTAAGACAAGCCCGTAACAGACAATATACCGTGGAATTATTATGGGGCCCCTA GCCGCTAACTTAGCTTGTTTGAAAAGTCTGCCCACCGTAATGGAGGGCCCCACTCAAAGTGAAAGAAGAGGTCCCTCGCGGCTACTCTACCTCGAGCCCCGAAGAAGCC TATTACGCATGGTCGCCACCAAGACAACAGGGTCTTATAGGGGCACGCTAAAAGAAGTATTATCACATCGAAACAATCACGTCAACCCGCTCCGAACGGTGGAAAACAG CTCAACATTGACTATATCTGAACAAGACTACTGATACGCGTCAGTGTCCAAGCTGTGGAAGCAATCACACATGTGGCCCGCAACCCAAGTTTTAAGAATTCTCGTTTCA' )"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "T7-TWHNK-P8T"
      },
      "source": [
        "# Testing my sanity.\n",
        "print( 'TTACCTTAAC GATATCTGTC ACGGCGTTCG CCCTAAAGAG CGTCAGAGGT'.split() )"
      ],
      "execution_count": 1,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "p8ZrruhB-cK8"
      },
      "source": [
        "# A motif scoring function."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "i-vWVl94-RTM"
      },
      "source": [
        "def Score_Motif( motifs, consensus_string ):\n",
        "\n",
        "  score_motif = 0\n",
        "\n",
        "  for kmer in motifs:\n",
        "\n",
        "    for i in range( len( kmer ) ):\n",
        "\n",
        "      if kmer[i] != consensus_string[i]:\n",
        "\n",
        "        score_motif += 1\n",
        "\n",
        "  return score_motif"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "oRlpBxAS-l5b"
      },
      "source": [
        "For this challenge I need to generate a random motif, then find the profile and then find a motif from the profile. Continue this until the score of the motif no longer improves. \n",
        "\n",
        "I do the above 1,000 times and then choose the one with the lowest score."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Vhe-ysXm-kxE"
      },
      "source": [
        "import random\n",
        "\n",
        "\n",
        "alphabet = [ \"A\", \"C\", \"G\", \"T\"]\n",
        "alpha_dict = { \"A\": 0, \"C\": 1, \"G\": 2, \"T\": 3 }\n",
        "motif_score = 0\n",
        "best_score = 0\n",
        "\n",
        "\n",
        "def randomizedMotifSearch( k, n, DNA ):\n",
        "\n",
        "  DNA = DNA.split()\n",
        "  genome_length = len( DNA )\n",
        "  motif = []\n",
        "  best_motif = []\n",
        "  kmer = \"\"\n",
        "  first_motif = True\n",
        "  print( 'This is running ')\n",
        "\n",
        "  for i in range( genome_length ):\n",
        "\n",
        "    # Selecting the DNA string to work on.\n",
        "    string = DNA[ i ]\n",
        "    # Determining the kmer being extracted from the DNA string.\n",
        "    rand_num = random.randint( 0, len( string ) - k  )\n",
        "\n",
        "    # Appending the sleected kmer to the motif array.\n",
        "    motif.append( string[ rand_num:(rand_num + k) ] )\n",
        "\n",
        "\n",
        "  # n is the number of iterations.\n",
        "  for j in range( n ):\n",
        "\n",
        "    #print( motif )\n",
        "\n",
        "    # Generate a blank profile matrix.\n",
        "    profile = [ ]\n",
        "\n",
        "    consensus_string = \"\"\n",
        "\n",
        "    consensus_prob = []\n",
        "\n",
        "    for nuc in alphabet:\n",
        "\n",
        "      blank_kmer = [ 1 for i in range( k ) ]\n",
        "\n",
        "      for mot in motif:\n",
        "\n",
        "        #print( 'mot: ', mot, 'motif: ', motif, 'k: ', k )\n",
        "\n",
        "        #print( 'motif: ', motif )\n",
        "\n",
        "        for j in range( k ):\n",
        "\n",
        "          #print( mot, j )\n",
        "          if list( mot )[ j ] == nuc:\n",
        "            \n",
        "\n",
        "            blank_kmer[ j ] += 1\n",
        "\n",
        "      profile.append( blank_kmer )\n",
        "\n",
        "\n",
        "    for i in range( k ):\n",
        "\n",
        "      total_freq = profile[ 0 ][ i ] + profile[ 1 ][ i ] + profile[ 2 ][ i ] + profile[ 3 ][ i ]\n",
        "\n",
        "      for j in range( 4 ):\n",
        "\n",
        "        profile[ j ][ i ] = profile[ j ][ i ] / total_freq\n",
        "\n",
        "        consensus_prob.append( profile[ j ][ i ] )\n",
        "\n",
        "      consensus_string += list(alpha_dict.keys())[consensus_prob.index( max( consensus_prob ) )]\n",
        "      consensus_prob = []\n",
        "\n",
        "\n",
        "    #print( 'profile: ', profile )\n",
        "\n",
        "    motif_score = Score_Motif( motif, consensus_string )\n",
        "\n",
        "    if first_motif == True:\n",
        "      best_score = motif_score\n",
        "      first_motif = False\n",
        "\n",
        "\n",
        "    elif motif_score < best_score:\n",
        "\n",
        "      best_motif = motif.copy()\n",
        "\n",
        "    for l in range( len( DNA ) ):\n",
        "      \n",
        "      for m in range( 0, len( DNA[ l ] ) - k + 1 ):\n",
        "\n",
        "        freq_table = {}\n",
        "\n",
        "        kmer = DNA[ l ] [ m:(m + k) ]\n",
        "\n",
        "        prob = 1\n",
        "\n",
        "        for p in range( k ):\n",
        "\n",
        "          # There may be an error here.\n",
        "          prob *= profile[ alpha_dict[ kmer[ p ] ]  ] [ p ]\n",
        "\n",
        "        freq_table[ kmer ] = 1 - prob\n",
        "\n",
        "      motif[ l ] = min( freq_table, key=freq_table.get )\n",
        "\n",
        "    #print( motif )\n",
        "\n",
        "  # Outputting the final resulting motif.\n",
        "  for kmer in best_motif:\n",
        "    #print('This is also running')\n",
        "    print( kmer )\n",
        "\n",
        "  return best_motif\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "TG05oX7t-rnD"
      },
      "source": [
        "best_motif = randomizedMotifSearch( 8, 1000, 'CGCCCCTCTCGGGGGTGTTCAGTAAACGGCCA GGGCGAGGTATGTGTAAGTGCCAAGGTGCCAG TAGTACCGAGACCGAAAGAAGTATACAGGCGT TAGATCAAGTTTCAGGTGCACGTCGGTGAACC AATCCACCAGCTCCACGTGCAATGTTGGCCTA')"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "DzxHT9p4-sD7"
      },
      "source": [
        "print( best_motif )"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "yDhEGQ5X-vWj"
      },
      "source": [
        "# A variant of the randomized motif searching algorithm."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ZU4-E7ct-xs6"
      },
      "source": [
        "\n",
        "import random\n",
        "\n",
        "\n",
        "\n",
        "alphabet = [ \"A\", \"C\", \"G\", \"T\"]\n",
        "alpha_dict = { \"A\": 0, \"C\": 1, \"G\": 2, \"T\": 3 }\n",
        "motif_score = 0\n",
        "best_score = 0\n",
        "overall_best_score = 0\n",
        "overall_best_motif = []\n",
        "\n",
        "random.seed(0)\n",
        "\n",
        "def randomizedMotifSearch( k, n, DNA ):\n",
        "    \n",
        "  random.seed(0)\n",
        "  \n",
        "  alphabet = [ \"A\", \"C\", \"G\", \"T\"]\n",
        "  alpha_dict = { \"A\": 0, \"C\": 1, \"G\": 2, \"T\": 3 }\n",
        "  motif_score = 0\n",
        "  best_score = 0\n",
        "  #overall_best_score = 0\n",
        "  overall_best_motif = []\n",
        "\n",
        "  DNA = DNA.split()\n",
        "  genome_length = len( DNA )\n",
        "  motif = []\n",
        "  best_motif = []\n",
        "  kmer = \"\"\n",
        "  first_motif = True\n",
        "  first_run = True\n",
        "\n",
        "  for i in range( n ):\n",
        "\n",
        "    motif = []\n",
        "\n",
        "    # Randomly generating a motif for the initial starting motif.\n",
        "    for j in range( genome_length ):\n",
        "\n",
        "      # Selecting the DNA string to work on.\n",
        "      string = DNA[ j ]\n",
        "      # Determining the kmer being extracted from the DNA string.\n",
        "      rand_num = random.randint( 0, len( string ) - k  )\n",
        "\n",
        "      # Appending the sleected kmer to the motif array.\n",
        "      motif.append( string[ rand_num:(rand_num + k) ] )\n",
        "\n",
        "    first_motif = True\n",
        "\n",
        "    while True:\n",
        "\n",
        "      # Generate a blank profile matrix.\n",
        "      profile = [ ]\n",
        "\n",
        "      consensus_string = \"\"\n",
        "\n",
        "      consensus_prob = []\n",
        "\n",
        "      for nuc in alphabet:\n",
        "\n",
        "        blank_kmer = [ 1 for i in range( k ) ]\n",
        "\n",
        "        for mot in motif:\n",
        "\n",
        "          #print( 'mot: ', mot, 'motif: ', motif, 'k: ', k )\n",
        "\n",
        "          #print( 'motif: ', motif )\n",
        "\n",
        "          for j in range( k ):\n",
        "\n",
        "            #print( mot, j )\n",
        "            if list( mot )[ j ] == nuc:\n",
        "\n",
        "              blank_kmer[ j ] += 1\n",
        "\n",
        "        profile.append( blank_kmer )\n",
        "\n",
        "\n",
        "      for i in range( k ):\n",
        "\n",
        "        total_freq = profile[ 0 ][ i ] + profile[ 1 ][ i ] + profile[ 2 ][ i ] + profile[ 3 ][ i ]\n",
        "\n",
        "        for j in range( 4 ):\n",
        "\n",
        "          profile[ j ][ i ] = profile[ j ][ i ] / total_freq\n",
        "\n",
        "          consensus_prob.append( profile[ j ][ i ] )\n",
        "\n",
        "        consensus_string += list(alpha_dict.keys())[consensus_prob.index( max( consensus_prob ) )]\n",
        "        consensus_prob = []\n",
        "\n",
        "\n",
        "      #print( 'profile: ', profile )\n",
        "\n",
        "      motif_score = Score_Motif( motif, consensus_string )\n",
        "\n",
        "      if first_run == True:\n",
        "        overall_best_score = motif_score\n",
        "        first_run = False\n",
        "\n",
        "      if first_motif == True:\n",
        "        best_score = motif_score\n",
        "        first_motif = False\n",
        "\n",
        "      elif motif_score < best_score:\n",
        "\n",
        "        best_motif = motif.copy()\n",
        "        best_score = motif_score\n",
        "\n",
        "      else:\n",
        "        \n",
        "        break\n",
        "\n",
        "      #print( overall_best_score, motif_score )\n",
        "\n",
        "      if motif_score < overall_best_score:\n",
        "\n",
        "        overall_best_score = motif_score\n",
        "        overall_best_motif = motif.copy()\n",
        "\n",
        "      #print( profile, motif )\n",
        "\n",
        "      for l in range( len( DNA ) ):\n",
        "        \n",
        "        freq_table = {}\n",
        "\n",
        "\n",
        "        for m in range( 0, len( DNA[ l ] ) - k + 1 ):\n",
        "\n",
        "          kmer = DNA[ l ] [ m:(m + k) ]\n",
        "          #print( kmer )\n",
        "\n",
        "          prob = 1\n",
        "\n",
        "          for p in range( k ):\n",
        "\n",
        "            # There may be an error here.\n",
        "            prob *= profile[ alpha_dict[ kmer[ p ] ]  ] [ p ]\n",
        "\n",
        "          freq_table[ kmer ] = 1 - prob\n",
        "\n",
        "\n",
        "        motif[ l ] = min( freq_table, key=freq_table.get )\n",
        "        #print( motif )\n",
        "\n",
        "    \n",
        "\n",
        "\n",
        "  # Outputting the final resulting motif.\n",
        "  blank = \"\"\n",
        "\n",
        "  for kmer in overall_best_motif:\n",
        "    #print('This is also running')\n",
        "    #print( kmer )\n",
        "\n",
        "    blank += \" \" + kmer\n",
        "\n",
        "  print(blank)\n",
        "\n",
        "\n",
        "  return overall_best_motif"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "QRrJnBOV-zh7"
      },
      "source": [
        "random.seed(0)\n",
        "best_motif = randomizedMotifSearch( 15, 1000, 'CGAAATCTAAAACAAGGCGGAAGGAAATCAAAGGCCAACCCTTCTCGGATACAAGTTAGTGAAGTCGGGGCATGTTGGCGGTAGGCGTAACACGCGGGGTCTCGCTGTTGGAGGGAAAACTCGTATTACTGGTTACGACCCTCCTAAATCAATGTCACCGGCCGGTCACCGGGAACAGCGCCTGCGAAATCTAAAACAA GGCGGAAGGAAATCAAAGGCCAACCCTTCTCGGATACAAGTTAGTGAAGTCGGGGCATGTTGGCGGTAGGCGTAACACGCGGGGTCTCGCTGTTGGAGGGAAAACTCGTATTACTGGTTACGACCCTCCTAAATCAATGTCACCGGCCGGTCTATTCACGTGCTTCAACCGGGAACAGCGCCTGCGAAATCTAAAACAA ACATCCGAGCGTGCTTCATGCGTCTTTCGCTCAGGACAGAGGATCTTGGCTCAGGCGCTGACCCCGCTCTCTTTTGCTGGGCACAGAGCGGTTCGGGATCAGTGCTCCGTTGTATATTACGACAACATCTAACTTTGTACATAGCGCGGGCTGAAGGTGTGACATCCCAAGACGCGCGTACTTGGTCATCGATATGTGG ATTGAATGCTCCACATCTAGAGGGACGCGAGTAACAATTCGAGTTAGAAGGAGAATTCCCCTGTCTCACATCAGGCTTCACTTCGCGCATTGTTAATCGAGCCGATTAAGCGTGCTTCCAGTTATAGGTCACTAACCCGGGCACCCATCCGTTCTTAGGGGACACCTCTGTAGACAACAAGCGTCATGGACGCTAACTA TAAGAGGGAGGCTCCTCGCAAGAAGTGAAATATTCAGGCGGCTGATGACAGCCGCGCGAACGTCTTTCTGTATCGTGATTAGGTAACGGCCAGTTGTCCGACGTACCGCCTCCCATTCTACAGGTTGCCGGGAAGTGTAAAGCACCTTAATATGACGTTAATATGGGTGACTCCGGAGGATATATCGGTGCTTCATCCG CAATGGAAAGCACCCACACAAAGTAATATAAGCGTGGAGCAAGCGGAGTGTATACTTTTTAAACCGTAAATCAAATAGACTACGCAGTCGTAATAGGTAAACGGAGGCGTGAAAGCAGATTCCAAATCAACCACGCGGGTTCTCTCTGATAGCAGTTAGTAGTGCTCCCCAATTTGCTAATTTCCGTAAGAGCTGTTAA AATCATCACTAGTAGATTAGAACGAGGAGTTATAACTTTGCTTCACATTCATTAAAATTGTGAAGGATGTTACTCGTCGGTCCCCCTCTCGACTTCTCAGAATAAGTGACTCCTGGATGACCAGATTACATGTTCACGTCGCGTAACACCGTTCTGCGGCATACGTGGCGTTCCGACAGTCGGTTGAATGGTCCATGCG AGAAAGCTGCTCCAATTACCTGATTAGCCGTCGATGGGATCCCGGAAGCTACCGAGCCACACTGGGTTTGTAACAGGTCAGCAAGACAAGTCAATTACCCAGGTCGAGCTGAATCGATATCCCCGTGCTTCAGGGAGACCAATGGGGACGTTAGGACCACGCGACCCGAAGGAAAGGGATACCGTTGGATATGCTAGCT CCACATGCTAACGCAAAATAGCCACGAGGCCGTAGGATGGAGTTTCCTAGAGACCTTGGTAGCACCTCGTCATTAGTTGACAATCTTAGGTAGCAGTATCATCAGCCGTGAGCGGGGGCTGCGCGTTCCACTTTTAGACTTGTACATCCGAAAGCAAAGCGTGCTTCATGGACCTGATAGCGCGTCAGTAGTGTTCTGG TTTCACATGACTCCATCAGTTCTGGAGCTGGTGGAGTACTGGTTCGTGGCGAGCTTGTCAGTCGCGTCACAAGTATAAGCGTGCAAAATTCTCCACTGCGCTGTAGAGTAAACAGAAGGCCGTATTAAACAGAAGTAATTGAGAAGATGTGGCTCTTGGAGAGACAACCGCTCTTAGCGACCAGTCTAGCGAATGAGAT GCGACAATCACCCAAAACGTAACCTAACCGTCAAACGTTATAAGCCGCCTTCAGACTAACGGCTAAGTTGGGCCAGTACCAGCTTAAAGTCGACCGTCACCATTCCAATGCACATCTTTTCGCTCAAGCGACAGGTCTGTCGAGCTGCATGTGGCTTTTGTGCTCGTATCAGAACGTCGGGCCGACTCGGCTTTGTGAC TAGTTTGCATTTATATCCTGGCACACGCTGTTCAGGCTGTAAAATTTGATTGGTTGTCAGCATACTGTCCGTGCATTAACGGACTGTTTATACCGGCGTGCTTCATTGTGTACTATGTGCCATAATTGTAGTTTTAAGCCCCTAGGCTAACATGTCGTCTAAAGCGGTGTCATGGGTTGTACACGGTATTGAGAGTATT CGGAGGCAGACGGGAAACGCTCCTAGGTCGGGCTGAAGTGGGATGACGAATCAGAGACAATATAAGCGTAAATCAAAGTCTGTTTGGGCGTGATTTATGTAACGCTTCGAAAACCTGCCATGGGAACCTAGAGAAGCAACTCTAACCCCAGTCGATTTAACCCGTGGTAGGCAAGTTCATTGGCCGAAGAGCAAGTGTG ATCCAGTCCCAATCTGCTGTGCTGAAAAGTTTGAGCGAGATGACGGGTAAGGGTCGTCTAAGACCATCGGGAATGTACAGAGGCTCGCCCCTAAGCATCAGTCAAAGGTTATGCACCAGTTCATCCCTCGTGAAAGTATAAGCGGCTTTCAAGAGTGCTAAGTGATCACCTGAAAGTGCAGACGGCCGGCTAGCCGCCT ATGATTTCGATATTATAAGAGTCGGATTATATTGGGCAGTTCAGCGCGTGTAGGTTGGGGTTGGATTTGTCCTGCTACGCCATGCTAACACTATTTTTCGGTATTTCCAGAAAAGTACGCGCTCCATACTTCCTATAACATTGCTTCACACTCTTTTGCGCCAAAACTCCAAATAGTCGTACAAGTGAACAGAATTCAT AACAATATCGATCCTCGAATGCTACGTCCTTTCCCACATTTGACCTCCTTCTCCAACCGACGCGCAATTTTCATAAGCGTGCTTGGTTAAATTCCGTCGGCTGATACTACTAGCGGACGGCGAGGTATGGCACAGGGTAGTACCGATTAGGGTATTAGATGGAGACCTTTACTCGTAACCGTCCCACTTTGAGCCCATG CGGCCGAGAACAAAACTGCAAATGATGGCGCGCTCCAGGAAGGAGTTATACACGACCTGTAACCATCAGTGCACAGGTTCAACCAGGCCATTATTGTAGGGTGTCTTCATCGTGATTCTAAATGGCGATCGTTGCGGTTATTCGGTGGTAATGCTTCCGCGGTAGCTTATAGCTGTGCTTCAAACGTAAGTGACCCAAG GTTAGGCTAGATCCGGGAAGATCTAACGAAAAGAAGTCTCGTGACATCGTGGTTTGCGGCTCGCTTCGGTCATTATTAGATGGCAACTCGCTCCTAAAACCTAGGTGCGATGACGACCACGACGACATCTCAGGAACCCCTCCGAATGGCTTGCTAGGGGAGATCGCCTATAAGGTAGCTTCACAAGTCCTCAGTCTCG GTTGATTCATAGGCGTGTCCGCCGACGGTCAGCGATTCACATCATATAGGAAACGATACGGCCGGGCATTCTAACCGGGGTATAAGGTCGCTTCATCGACAATTGGTCGGGTATGTAAGTGATCGCTTTATACTCTCAATCGTCACTTGTGCCTTTTGGGCAATCTAGTATCGCAAAACCTAGCGGACGGCCCAATAAA GGTTGCCCCTCCTTATGACTCCAGCAACATATACGTATAAGCGTGCTGTCGTGCCGTCAAGCTTAATACGCTGAGCGGATCGATCGAAATTAGAACACCACCGTCAGGCATTTTCCGGGTGCTACACGTACTGTTAGGCTAAAACTATTCTATCAGCGTACGTTAGAAGCGCGATTCTAGTAGCAATGCTTGGTAAGCC')"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Sxpyh1Er-1nL"
      },
      "source": [
        "print( best_motif )"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "VFiFPPNm-4KL"
      },
      "source": [
        "# Outputting the final resulting motif.\n",
        "for kmer in best_motif:\n",
        "  #print('This is also running')\n",
        "  print( kmer )\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "WKvjt79J-8G7"
      },
      "source": [
        "# Gibbs sampler and randomized motif finder."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "n5Atwv9h_AK7"
      },
      "source": [
        "\n",
        "\n",
        "import random\n",
        "\n",
        "\n",
        "\n",
        "alphabet = [ \"A\", \"C\", \"G\", \"T\"]\n",
        "alpha_dict = { \"A\": 0, \"C\": 1, \"G\": 2, \"T\": 3 }\n",
        "motif_score = 0\n",
        "best_score = 0\n",
        "overall_best_score = 0\n",
        "overall_best_motif = []\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "def FindProfile( motif, k ):\n",
        "\n",
        "  # Generate a blank profile matrix.\n",
        "  profile = [ ]\n",
        "\n",
        "  consensus_string = \"\"\n",
        "\n",
        "  consensus_prob = []\n",
        "\n",
        "  for nuc in alphabet:\n",
        "\n",
        "    blank_kmer = [ 1 for i in range( k ) ]\n",
        "\n",
        "    for mot in motif:\n",
        "\n",
        "      for j in range( k ):\n",
        "\n",
        "        #print( mot, j )\n",
        "        if list( mot )[ j ] == nuc:\n",
        "\n",
        "          blank_kmer[ j ] += 1\n",
        "\n",
        "    profile.append( blank_kmer )\n",
        "\n",
        "\n",
        "  for i in range( k ):\n",
        "\n",
        "    total_freq = profile[ 0 ][ i ] + profile[ 1 ][ i ] + profile[ 2 ][ i ] + profile[ 3 ][ i ]\n",
        "\n",
        "    for j in range( 4 ):\n",
        "\n",
        "      profile[ j ][ i ] = profile[ j ][ i ] / total_freq\n",
        "\n",
        "      consensus_prob.append( profile[ j ][ i ] )\n",
        "\n",
        "    consensus_string += list(alpha_dict.keys())[consensus_prob.index( max( consensus_prob ) )]\n",
        "    consensus_prob = []\n",
        "\n",
        "  return profile, consensus_string\n",
        "\n",
        "\n",
        "\n",
        "def Score_Motif( motifs, consensus_string ):\n",
        "\n",
        "  score_motif = 0\n",
        "\n",
        "  for kmer in motifs:\n",
        "\n",
        "    for i in range( len( kmer ) ):\n",
        "\n",
        "      if kmer[i] != consensus_string[i]:\n",
        "\n",
        "        score_motif += 1\n",
        "\n",
        "\n",
        "  #print( score_motif )\n",
        "  #print( len( motifs ) )\n",
        "\n",
        "  return score_motif    \n",
        "\n",
        "\n",
        "def Gibbs( k, N, n, DNA ):\n",
        "\n",
        "  alphabet = [ \"A\", \"C\", \"G\", \"T\"]\n",
        "  alpha_dict = { \"A\": 0, \"C\": 1, \"G\": 2, \"T\": 3 }\n",
        "  motif_score = 0\n",
        "  best_score = 0\n",
        "  overall_best_score = 0\n",
        "  best_motif = []\n",
        "  overall_best_motif = []\n",
        "  motif = []\n",
        "\n",
        "  DNA = DNA.split()\n",
        "\n",
        "  genome_length = len( DNA )\n",
        "  first_motif = True\n",
        "  first_run = True\n",
        "\n",
        "  for j in range( n ):\n",
        "\n",
        "    motif = []\n",
        "\n",
        "    # Randomly select a motif.\n",
        "    for i in range( genome_length ):\n",
        "\n",
        "      string = DNA[i]\n",
        "      \n",
        "      rand_num = random.randint( 0, len(string) - k )\n",
        "\n",
        "      motif.append(  string[ rand_num:(rand_num + k) ] )\n",
        "\n",
        "    # Running through N times.\n",
        "    for run in range( N ):\n",
        "\n",
        "      # Selecting the string to optimise.\n",
        "      j = random.randint( 0, genome_length - 1 )\n",
        "\n",
        "      motif_sans_j = motif[ :j ] + motif[ (j + 1): ]\n",
        "\n",
        "      profile, consensus_string = FindProfile( motif_sans_j, k )\n",
        "\n",
        "      freq_table = {}\n",
        "\n",
        "      # Finding the profile most-probable kmer within the jth string.\n",
        "      for m in range( 0, len( DNA[ j ] ) - k + 1 ):\n",
        "\n",
        "        kmer = DNA[ j ] [ m:(m + k) ]\n",
        "        #print( kmer )\n",
        "\n",
        "        prob = 1\n",
        "\n",
        "        for p in range( k ):\n",
        "\n",
        "          # There may be an error here.\n",
        "          prob *= profile[ alpha_dict[ kmer[ p ] ]  ] [ p ]\n",
        "\n",
        "        freq_table[ kmer ] = 1 - prob\n",
        "\n",
        "      #motif[ l ] = min( freq_table, key=freq_table.get )\n",
        "      motif_sans_j.insert( j, min( freq_table, key=freq_table.get ) )\n",
        "\n",
        "      motif = motif_sans_j.copy()\n",
        "\n",
        "      motif_score = Score_Motif( motif, consensus_string )\n",
        "\n",
        "\n",
        "      if first_run == True:\n",
        "\n",
        "        overall_best_score = motif_score\n",
        "        first_run = False\n",
        "\n",
        "\n",
        "\n",
        "      if first_motif == True:\n",
        "        best_score = motif_score\n",
        "        best_motif = motif.copy()\n",
        "        first_motif = False\n",
        "\n",
        "      elif motif_score < best_score:\n",
        "\n",
        "        best_motif = motif.copy()\n",
        "        best_score = motif_score\n",
        "\n",
        "      #print( motif_score, best_score, overall_best_score)\n",
        "\n",
        "    if first_run == True:\n",
        "\n",
        "      overall_best_score = best_score\n",
        "      first_run = False\n",
        "\n",
        "    if best_score < overall_best_score:\n",
        "\n",
        "      overall_best_motif = best_score\n",
        "      overall_best_motif = best_motif.copy()\n",
        "\n",
        "    #print( overall_best_score, motif_score )\n",
        "\n",
        "      \n",
        "  # Outputting the final resulting motif.\n",
        "  blank = \"\"\n",
        "\n",
        "  for kmer in overall_best_motif:\n",
        "    #print('This is also running')\n",
        "    #print( kmer )\n",
        "\n",
        "    blank += \" \" + kmer\n",
        "\n",
        "  print(blank)\n",
        "\n",
        "\n",
        "\n",
        "  return overall_best_motif\n",
        "\n",
        "    \n",
        "\n",
        "\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ADWk6wrC_DUr"
      },
      "source": [
        "best_motif = Gibbs( 15, 2000, 50, \"ACCTACGACTAAATAGTACCTCCTACGTGAGCCCGCGGCGCACCCCGATAAATACCGTCGGCTTGGAGACTCGGGCCCAATGCTGGTTAGATCTACAGAGCGGAACACTGCACCATTACATCGCCCGCATACGCTCTACTGAGGCACCACCACGTGATGCACACACGGACGCGTCGCTAGAGTTTTGCAAACACGAGGAACGTAAAGAGAGAGTGAGGACTTCCCGGTATCGAGAGACGCGGATCTGATAGTCACGACATATCAGTGCAAATTGGTGCCTGGAGCGCATTTACTGCTGTCCGTAGTACACATAGGACCTACGACTAAATA GTACCTCCTACGTGAGCCCGCGGCGCACCCCGATAAATACCGTCGGCTTGGAGACTCGGGCCCAATGCTGGTTAGATCTACAGAGCGGAACACTGCACCACTTAATTGCTGAACATTACATCGCCCGCATACGCTCTACTGAGGCACCACCACGTGATGCACACACGGACGCGTCGCTAGAGTTTTGCAAACACGAGGAACGTAAAGAGAGAGTGAGGACTTCCCGGTATCGAGAGACGCGGATCTGATAGTCACGACATATCAGTGCAAATTGGTGCCTGGAGCGCATTTACTGCTGTCCGTAGTACACATAGGACCTACGACTAAATA CCTTTAATAGGACCTGGAGGCTTGACCCAGAATCTAGGATTTTGCTACTCCGAAAACCATCCTACCCGTGCAACCGCACGGGACTACGGATCGGCGACGAACCAATTCGAACCGGAAAGTATGGTAGGTCTCCTCTACCGCGATTAAATTCTCTCAGTCGCGCTGCTGATCAAGACACACCCACGTGCTGAATAGTCAGGCGTCCAAGGTCGTGAAGCCATTAGGACTTCGTGTGTCCGGAGCTCGGCAAGCTCACGTGTTTTTGCGGGCGTTCCGCCTTCTACCACTCTCCAAAGTTTCTTCTTTTCCTTTGAAGTTTGTCATACAGCG GGAACCTACGTACCCGGTTATTTAATAGTCGACCCGGAGAGGGGTTTAATTGCACCAACAAACTACGTTATGGCTAATTCGGGAGACCTATACACCCTATTGTGAAAGAATCCTGTACCAAGTTGCGTTGTTCAAGCTGGAGGAGCTAACCTCTGACCGCATCGTTCGAAGTCAGCTGCCGATTATCTTTAAGGAGAGAGTGGGGGCGTTCGCCCCCCGGGTTAAAAATGCGATCATTTACCCTAGCAATCCGGTATTGAGCGGCAGGTGACCTACTTTAGAGGCGGAGCTCATAATTACACTGACATCCAACTTGGTAGAATGGGCTCC CATAAATAAACGATGCGGGGCCTGATTAGGCCTAGGTGACGCACACTATTTCATGCATGCCTTATATCGTTAATTATGAGATGCTTTGTAAATGATGTACAACACCGACCAGGGCCCTCTTGTAAAGATGAACAAGATCGGAAGGCGACCAAAATCGCAATAAGCACCGCCTTGCTGAATGAAAGATGCCAAGAGCTTGCCCAAACAGTGTTCCGCCTGCGATGTTGGGGGTTAATCAGATAGCGAGCCCAAGCATTCTAAACCGAATAAAGACAGAGGCTTTACTTGCATAGACTAAGAGTAAAGGTTGATACATACGTCGTTGAAAGG GTCCTTAGTCCGCCTTAGCCCCCTCAAGGGAATAATACAAGTCCCACCAATCATGCGCTTGCCCCATTCCCCATGAGGGGATTGGATCAATCTCCCGGAATAGAGAAAGTTCCTCAAGGTATTGCTGAACATCGTCCTTAATAGTCACGTGCGTCAGTGGGGAAGGGGCTCATACCGGCCCAATGGTAGAAGCCAGTCGTTTTTGAGCGGAGGGATCCGTCTTCGCAACTCTCCAATGGCTAAAGCCCGTAACACAAGCCCGGCTGGCCCCTTGCGCCTTAGTTCAGTGCCCGGGGAGCAGGTGTCTCTGTACGGCTGCACTCTCGGACG TTGCTGCACATCTTGGCGGCGTTTAGCCCGACGACCGCCACCCGTGGACTGATATCCTTACGCCGTATGTCAGTGACTGTTGAGTCAGGTCTTTACAGGAGGCCGTAAGTTGAAGATTTAAGGCTCTTCGCTAACTGTCCGTAATTCGAGTGCAAGCCATAAGGTTTGGGTGGTTGCCCTATTGCTGACTAGACGTGGCATCCAGCAACAGAGCCGCATTATCGACCTGAAACGGGCATTCTCCTCTCAGTTAATCATGCGGACTGCACGTGCGGTAGCCCCGTCGAAAGTTGTCCTTGCGACCAAGTGGATATGGATTTTTATTATTCT CTGTCCTAATTTTTATTGAGTACAGGCATGGCGTTTGATTTCCTTTCCAATGACGGATGTTACGAGCTATGTAAGCTACCCGGCTGCAAGGATGGATGGCAGTCGGTGGTCACTTTACTCGCGTCTGTTCTGAGGCCTTCATATAGCATGTCACGTGACAACCTCTTTACACACGCGAGAGAATGTTTAGGGAATTCCTTCGCATAAATAAACTGCCCTGTCCCTTTAGTTGAGCTTATGTGATGCTGGCACTGACGGGAGAAGTGCGTCCATTTACTCATGGCCCACCCTATTGCGCTATTAATGGTTCAATAGTACATTTGTAAGGTT AAGCGATATATGAACGAGAGCTTGCGCGACAAGTCGGGTCCGGCCTTTGGGGGAGTTAGCGAGCGGCACCCTACGCCTGAATTTTGAGGCCTCGCATACAACATTTTGAACCCTCGGACGTGTCGGGCTCAGATCCACAAAGAGGGATAGCATTGTGAGGTTAGTAAGGCTGTTCGACTAGACATTATGACCGTTGGAACTGTTTTCGGAGTGCTGCAATCCCGGATCATACAGAGCTAACGGGCTCAGGGCATATTGGTGAAATCCCGCGCGGGCCCCTAAAGGCTTGCTGCACCGAAGACCGAGAGGGACCTTGGCTTCCCTCCGCTG CGCAGTCTGATGTCTGACGCGTGCTGGTCGGGGTGCCCAAAAAGCCGTCTCACTTGCTAGAGGTAATGGCCACCCTATCATTGAACCCTATGTCTGTCGTGAAAGGGACTAACATATACAAATGTAGTCGTTAAAGTCTCAGCCTCATACAAGTCTCCGAGGAGATGCTGTCATGTCCTACTATGCGTGCGCGCTTGAGTTCCATCCCGGGTCATTCTTGCGGGTGTGAGAGTAGAATTTGGAGTACGGGTCGGCTCAAACTGGTTATGCATAAATCGTTTGTTTTGTCATTTCTTGGAAGAAACGCAACTTCACGCAAGTGGGTGTAAT CCCTCCTGGACCAACAGCGAGTGAGCGTGTTAGTACGTCGGTGGGGAGAGCTTAGAGCCGGTCTGTATACCTTCATGCATTGTGCTTCGCTAAAGCATTTATAACAACGACCTTCGCACTTCCGCCGCTCCTTGGATAGCTAAAACAGGTCCGACCATCGCCAAGCTGTAGGCTCAAAGCTACAATACTTGAGCTTCCCTCGAATCTCCGTCCATTCCTTATAAAGACCTAGCGTACTCTGGTGAGTTACCGAGGAAAGGATTAACCAAACTTTGATATGACACACCCTTGGGCTGAATAACGCCCGGCGGTTTAGAATAAATAAGAAAT CCCTTCCTTAGGATCATCGCAATTCCGCCTGCACACCCTCCATAGAAGCCGAAAGCGCCAAGTTAAACAGCAGTTGCTACACGCTAGGCACAGTAACATCATAGATTTGAAATTCCTGGCGAGGTAAGGACTAGGTTAGGCACCAACCCTTAAGGTAGTGAAACGGAGAGAGTAGCCCGTCCATCACCATTGTTACGAAGGCAAACCTAGTCTAGGTTGTCTGTACCTGACCCCTGCCTTGCGCTAGCCCCGGGGGCAGCTATATCGTTTTTAACGACTGTCGATAGAGGCAAAATCAGCACTTCATTGCTGAAAACCTCGAAGGCAGCT GCAGCACGCACCCTATTGCTATTCTTCTCAGCATGGCTAAAGTGCTCGACCTATTCAATAAAACCCATATACGGATAAGATCCTGCGAGAGCAGGCCTTCCAACTAGAGGTGGTGTTTTATACACAGTTCCACCAGGGAAACGGGGGACAAGAGATACCATACATAGCTTTCGTTATCGAGTTTGACGGAAAACCACCCATTGTACATAGTATGACTGTTGTTCACTAAGGCTCCGAATATCGGATAACGATGGTTATCCGTTGTCACCGACCGCTTCGACCACCCGGCCCGAGTTTATGTATCTCCATCAGAGGGTACATACACATGCA GCATATCAATTGAGGGGCTCTTCGAAATCGCACACGGAGGATCTAACGGCCAGGCGTGTTTGTGTATTGCTGAGGTGGTCCACGTAATAAGTAAAAGATTCTCAAGGGGGACATGCGTAACCGAGGTTACCCCTTTAGGGGCCGATACTTGGCCGGCCCTAAGATATAGAGCTGTTTCCCTTAGCGTACATGCCATCCGTTAGTATTACTTACTGTATGCTTACCCTATTGCTGTCAAGTTCGTTTGAGTTCACAGGCCCTCTTTTGACAAGAATCAGAGGATTGTCCAACATGAAGCACCCGACGTCAGGTAAGCTGGTAGCAGTAAAT GCCGACGACTGCATGTTAGAATGTCTAACCTGACGACCAGCGACCGGGCCTCTCGTGCCTGTAGCCATGGGGGCCTTGAATATGGAGTCAGACAGCCCGCATGGTCGGATTCTCCTAGGGATGGGTCATCGAGGGTACCAATGTAAAGGATCCTGAAATGTTGGAAAACGCACCCTATTCGGGAAGTGACTCCCAGAAGGTAGCCGCAAACTACTACGATCTCCTTTACCGCCCACCCAGGATCCATATCACCTAGCTGAGGACGTGGGGCGAGACAGTATCAGGGTTAACTGATTCTCACGTTCGGTAGTGATATTAAGATATTTCACG CTCGTGCAAATGGCTAGTGAGTTTAGCTTCGGCGAAAGTGCACCGGAACTCCTTGGCCCCGATGTTTATCGAACCGCTATTAAGCACGACGACCAGACCCTTGTATCTGCGGCAGTTCCTCAAAAACCTCGATGACCAAATGTCACCCTCGAGCTGAATGACGTAGCTCCGGTCGTCACGACTCCGGGGATTTCGTTGTTTGCACGAGAATACGTTGTTATTAAGTAGCTGTGAACTCTGCCAACACAGCCGAAGTTTATATGACGGAGAGGGGACGCCTCTAGCGTTACTATGGCCCATTTAAGGGATATCGGACGTGTAGTCTCACCG CGGTACAGTGACGCTCCTATTGCTGAAATTCGATGCGCCGCACGCATATCCCAGCCAAAGCGGTTGCTACATGAATTGCCCCGATATCTAGGAACGATGCGCTTCAAGTCGCGACTTTAGCTAAATCATAGATCCACGGAGGTTATAGACAGTTTTCACCGCCAAAGATTTTTATGTTACACCGCATTGGCCTGATCAATTGCCCTACCTGCAAATCCTCCTACGTACATTGTACAAATGCATCGGGTGGGCAGGCAATGCGACAACGACGAAGTAGGCTTAGTCATGATGATCCGGTAAGACATTTACTACAGGAACAAACAAGTGCGA TATGGATCGTGAGCGACGTTCTGTCCGAGTGCAGTGACGCCGCACTCTCCTGGACGCAGCAGCTGGCTCACCCATATGCTGAAGACTAGCGCGGTATCTCGTGGTAAGTTATCTATCATGCACGGTGGGTCTGCAATACTTCCGCCCACTATGGCGTGTTGAAGGGTTAGGTACGTGGACTCGGGTGGGATCCACACCGCCTCTCCATTCTAGGGTGCACCAGCTCGTTACCGGCTGCTCCAGTGTGGGGACTCCCGTCCGATCAATAAGCCCAATAGAAGAAGGTTGATTGCGTGGTCTGATACCCCACGTAAACACAACGCTGCCTTC CTACGCTGGAGTCGTAGAGGCCTTATGCTCTCAGAAGGCGGGTTATGGATAGTCGCACCACTTACACATAGGCTACCGAACGAGGTTCACTGACGGGTACTGGCTTCTGCGTGCCGGTACGCCTATCGCGGGCTCTGCTCAGAAGGAATGGTGTCGTCGAGTTTATTCAAAGCGTTGAATGCAGTTGCACTTCCCAATCGAACCCAAAAAGCGACGAAGGACAGTTTAACCCAGGGTTGGTGTCCATAACTCTGCGACGATCCTCACGGTGAGCGGTATCGGAGTCACCGATTTGCTGAATGAGATTAAGTTTGGGACTTTGCTTCGCCT CGGGGTAGCGCTGATCTGGTTAACTACTAACCCCGAAAGCTCAGCCTAGGTCTATCTCATGTAATCGTTGACGTTTCTCTGTAGCACAGTGCCCCGATAGCTAAACTTGCTATTGCTGAAGTTGTTATAGATTATCGGTGTGACGGGTGAAAGATTGAGAGGGTCGCATTGTAAAGCATCATTCTGATACATACACGAGATTCCGGTTACACGAATACCCTTTTCACACCTTGACAAGTCCTATGCCGCGTGCCCTAGCCCTCTGGTGATCGGTCACGCACAGTGACAAGCCCAACATACCATATAAGTAACGCCCTAGTTTTATTAAGT\" )"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "BadOmOpG_GRS"
      },
      "source": [
        "print( best_motif )"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "r0nPQY5R_Jhb"
      },
      "source": [
        "# Finding the Hamming distance between a gene and multiple strings of DNA."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "5cFjxBql_Owa"
      },
      "source": [
        "DNA = \"CTCGATGAGTAGGAAAGTAGTTTCACTGGGCGAACCACCCCGGCGCTAATCCTAGTGCCC GCAATCCTACCCGAGGCCACATATCAGTAGGAACTAGAACCACCACGGGTGGCTAGTTTC GGTGTTGAACCACGGGGTTAGTTTCATCTATTGTAGGAATCGGCTTCAAATCCTACACAG\"\n",
        "\n",
        "DNA = DNA.split()\n",
        "\n",
        "def HammingDistanceArray( fg, sg ):\n",
        "\n",
        "\n",
        "  distance = 0\n",
        "\n",
        "  for i in range( len( fg ) ):\n",
        "\n",
        "    for j in range( len(sg) ):\n",
        "\n",
        "      \n",
        "      if list( fg[ i ] ) [ j ] != sg[ j ]:\n",
        "        distance += 1\n",
        "\n",
        "  return distance\n",
        "\n",
        "\n",
        "HammingDistanceArray( DNA, \"TAGTTTC\" )\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}